src/Grpc.AspNetCore.Web/Internal/GrpcWebFeature.cs

```diff
@@ -0,0 +1,94 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+
+namespace Grpc.AspNetCore.Web.Internal
+{
+    internal class GrpcWebFeature : IRequestBodyPipeFeature, IHttpResponseBodyFeature, IHttpResponseTrailersFeature
+    {
+        private readonly IHttpResponseBodyFeature _initialResponseFeature;
+        private readonly IRequestBodyPipeFeature _initialRequestFeature;
+
+        private readonly Base64PipeReader? _pipeReader;
+        private readonly Base64PipeWriter? _pipeWriter;
+        private IHeaderDictionary _trailers;
+        private bool _isComplete;
+
+        public GrpcWebFeature(GrpcWebMode grpcWebMode, HttpContext httpContext)
+        {
+            _initialRequestFeature = httpContext.Features.Get<IRequestBodyPipeFeature>();
+            _initialResponseFeature = httpContext.Features.Get<IHttpResponseBodyFeature>();
+
+            if (grpcWebMode == GrpcWebMode.GrpcWebText)
+            {
+                _pipeReader = new Base64PipeReader(_initialRequestFeature.Reader);
+                _pipeWriter = new Base64PipeWriter(_initialResponseFeature.Writer);
+            }
+
+            _trailers = new HeaderDictionary();
+        }
+
+        public PipeReader Reader => _pipeReader ?? _initialRequestFeature.Reader;
+
+        public PipeWriter Writer => _pipeWriter ?? _initialResponseFeature.Writer;
+
+        public Stream Stream => _initialResponseFeature.Stream;
+
+        public IHeaderDictionary Trailers
+        {
+            get => _trailers;
+            set { _trailers = value; }
+        }
+
+        public async Task CompleteAsync()
+        {
+            // TODO(JamesNK): When CompleteAsync is called from another thread (e.g. deadline exceeded),
```

Comment:
As middleware we can't add locks to the server gRPC library.

Could locking inside PipeWriter work? e.g. lock to prevent new calls to GetSpan/GetMemory/Advance from happening? But that wouldn't help in situations where gRPC server was halfway through writing a message.

Reply:
You really need to re-think how you do timeouts then. Multi-threaded request/response access must be prevented. Calling CompleteAsync from a timer thread is fundamentally unsafe.

Reply:
A solution would be to add a service that is responsible for writing messages. gRPC-Web could then use DI to wrap it, and add locking to the wrapper to co-ordinate with CompleteAsync.

This will require changes to the core gRPC library so I'll leave it as is until we decide we want to commit to supporting this.